File: api/v1/endpoints/daily_reflections.py
# -*- coding: utf-8 -*-
"""“渡劫”每日心得端点。"""

from __future__ import annotations

from datetime import date
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query

from api.deps import require_permission
from api.v1.schemas.miniapp import (
    DailyReflectionDeleteResponse,
    DailyReflectionItem,
    DailyReflectionListResponse,
    DailyReflectionStatsResponse,
    DailyReflectionUpsertRequest,
)
from src.services.daily_reflection_service import (
    DailyReflectionError,
    DailyReflectionNotFoundError,
    DailyReflectionService,
)
from src.services.wechat_miniapp_auth_service import MiniappPrincipal

router = APIRouter()


def _not_found(exc: Exception) -> HTTPException:
    return HTTPException(status_code=404, detail=str(exc))


@router.get("", response_model=DailyReflectionListResponse, summary="列出当前用户心得")
def list_reflections(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    principal: MiniappPrincipal = Depends(require_permission('daily_reflections.read')),
) -> DailyReflectionListResponse:
    payload = DailyReflectionService().list(
        user_id=principal.user.id,
        page=page,
        page_size=page_size,
    )
    return DailyReflectionListResponse(**payload)


@router.get("/by-date/{reflection_date}", response_model=Optional[DailyReflectionItem], summary="按日期查询心得")
def get_reflection_by_date(
    reflection_date: date,
    principal: MiniappPrincipal = Depends(require_permission('daily_reflections.read')),
) -> Optional[DailyReflectionItem]:
    payload = DailyReflectionService().get_by_date(
        user_id=principal.user.id,
        reflection_date=reflection_date,
    )
    return DailyReflectionItem(**payload) if payload is not None else None


@router.put("", response_model=DailyReflectionItem, summary="保存当日心得")
def upsert_reflection(
    request: DailyReflectionUpsertRequest,
    principal: MiniappPrincipal = Depends(require_permission('daily_reflections.manage')),
) -> DailyReflectionItem:
    try:
        payload = DailyReflectionService().upsert(
            user_id=principal.user.id,
            reflection_date=request.reflection_date,
            title=request.title,
            content=request.content,
        )
        return DailyReflectionItem(**payload)
    except DailyReflectionError as exc:
        raise HTTPException(status_code=400, detail=str(exc))


@router.get("/stats", response_model=DailyReflectionStatsResponse, summary="连续打卡与月度回顾统计")
def get_reflection_stats(
    reference_date: Optional[date] = Query(None, description="客户端本地今天(YYYY-MM-DD)"),
    month: Optional[str] = Query(None, description="回顾月份(YYYY-MM)"),
    principal: MiniappPrincipal = Depends(require_permission('daily_reflections.read')),
) -> DailyReflectionStatsResponse:
    payload = DailyReflectionService().stats(
        user_id=principal.user.id,
        reference_date=reference_date,
        month=month,
    )
    return DailyReflectionStatsResponse(**payload)


@router.get("/{reflection_id}", response_model=DailyReflectionItem, summary="读取一条心得")
def get_reflection(
    reflection_id: int,
    principal: MiniappPrincipal = Depends(require_permission('daily_reflections.read')),
) -> DailyReflectionItem:
    try:
        payload = DailyReflectionService().get(
            user_id=principal.user.id,
            reflection_id=reflection_id,
        )
        return DailyReflectionItem(**payload)
    except DailyReflectionNotFoundError as exc:
        raise _not_found(exc)


@router.delete("/{reflection_id}", response_model=DailyReflectionDeleteResponse, summary="删除一条心得")
def delete_reflection(
    reflection_id: int,
    principal: MiniappPrincipal = Depends(require_permission('daily_reflections.manage')),
) -> DailyReflectionDeleteResponse:
    try:
        DailyReflectionService().delete(
            user_id=principal.user.id,
            reflection_id=reflection_id,
        )
        return DailyReflectionDeleteResponse(deleted=1)
    except DailyReflectionNotFoundError as exc:
        raise _not_found(exc)

File: apps/dsa-web/src/api/dailyReflections.ts
import apiClient from './index';

// “渡劫”每日心得。后端为 /api/v1/daily-reflections（中性前缀，Web 与小程序复用同一处理逻辑）。
export type DailyReflectionItem = {
  id: number;
  reflection_date: string;
  title: string;
  content: string;
  created_at?: string | null;
  updated_at?: string | null;
};

export type DailyReflectionList = {
  items: DailyReflectionItem[];
  total: number;
  page: number;
  page_size: number;
};

export type DailyReflectionStats = {
  total: number;
  current_streak: number;
  longest_streak: number;
  today_done: boolean;
  month: string;
  month_count: number;
  month_days: number[];
};

export const dailyReflectionsApi = {
  async list(page = 1, pageSize = 20): Promise<DailyReflectionList> {
    const { data } = await apiClient.get<DailyReflectionList>('/api/v1/daily-reflections', {
      params: { page, page_size: pageSize },
    });
    return data;
  },

  async getByDate(reflectionDate: string): Promise<DailyReflectionItem | null> {
    const { data } = await apiClient.get<DailyReflectionItem | null>(
      `/api/v1/daily-reflections/by-date/${reflectionDate}`,
    );
    return data;
  },

  async upsert(reflectionDate: string, content: string, title = ''): Promise<DailyReflectionItem> {
    const { data } = await apiClient.put<DailyReflectionItem>('/api/v1/daily-reflections', {
      reflection_date: reflectionDate,
      title,
      content,
    });
    return data;
  },

  async remove(id: number): Promise<{ deleted: number }> {
    const { data } = await apiClient.delete<{ deleted: number }>(`/api/v1/daily-reflections/${id}`);
    return data;
  },

  async stats(referenceDate?: string, month?: string): Promise<DailyReflectionStats> {
    const { data } = await apiClient.get<DailyReflectionStats>('/api/v1/daily-reflections/stats', {
      params: { reference_date: referenceDate, month },
    });
    return data;
  },
};
